### PR TITLE
Attrs - Initial progress

### PR DESCRIPTION
**Pull request by azarzadavila**
As discussed in #7 @PgBiel and @kilacoda started working on progressively ditching CONFIG to use instead `attrs`.
By conversing with PgBiel I thought I could give a bit of help. This PR is currently more a way to see and discuss how attrs could look like in the project.
I would also like to bring a few issues with using this approach.

## List of Changes
- Mobject using attrs
- geometry.py using attrs
- test_geometry.py compatible with the changes

## Issues with attrs
`attrs` is not really inheritance friendly. As manim abundantly uses inheritance this results in a few issues/shortcomings.
- **MAIN ISSUE** A class that inherits from a class that possesses a keyword argument for initialization can not have a positional argument.
This means for example that if `TipableVMobject` uses attrs and has a keyword argument than `Line` can not use `start` and `end` as positional arguments.
- One advantage of using attrs is to specify a type to each attribute. However if one wants to redefine the default value for a parent attribute, than the type of said attribute must be re-specified. For example, `Circle` that uses RED as default subclasses `Arc` that uses WHITE as default. Both classes must specify the type of arg color. It could look like this :
```python
class Arc(TipableVMobject):
    color: typing.Union[str, Color] = WHITE

class Circle(Arc):
    color: typing.Union[str, Color] = RED
```
Thus repeating `typing.Union[str, Color]`.

## Possible solutions
- `attrs` suggests the use of `@classmethod` for some initialization. 
This could be used for cases where the arguments in the master `__init__` version uses positional arguments and/or the passed arguments should not be attributes of the class. For example a method for `Line` could be included :
```python
    @classmethod
    def from_start_end(cls, start, end, **kwargs):
        return cls(start=start, end=end, **kwargs)
```
- a special validator could be used to indicate some initialization arguments can not be None.
This would still require the use of keyword arguments but it would ensure that they are passed anyway.
 
These solutions would still break code that includes positional arguments.

## Testing Status
It passes two tests :
- `test_creation.py`
- `test_geometry.py` which has been modified to remove positional arguments.

## TODO
- find a way to still keep CONFIG for backward compatibility reasons. This should not be too hard, I believe it is just a matter of placing an adequate function in `__attrs_post_init__`.
- choose adequate types for all attributes
- implement adequate validators
- further replace CONFIG in other parts of the code
## Further Comments
A custom decorator `dclass` was generated to not repeat what was chosen as parameters in the `attrs` decorator : 
```python
dclass = functools.partial(attr.s, auto_attribs=True, eq=False)
```
(slots=True will normally also be part of final implementation).
Custom decorators like this are not analyzed by PyCharm this means the attributes are not part of the autocompletion.
However `attr.s` is recognized by PyCharm (I don't know for other IDEs) so I think I'll prefer to rewrite each time the decorator.

I hope this will bring more insight on how `attrs` could be integrated to manim and I'll try to keep working on this PR this week.

## Acknowledgement
- [X] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

### CODE REVIEW

File: manim/animation/animation.py [hunk 1]
```diff
@@ -1,50 +1,17 @@
...
-from ..utils.dataclasses import dclass
-
-if typing.TYPE_CHECKING:
-    from ..scene.scene import Scene
 
 
 DEFAULT_ANIMATION_RUN_TIME = 1.0
 DEFAULT_ANIMATION_LAG_RATIO = 0
 
-@dclass
```
Comment (kilacoda) [inline manim/animation/animation.py:19]: Why revert everything we've done on `Animation` though?

**Comment by PgBiel** [comment]
First of all thanks for the PR
While I cannot review right now, may you please replace `@dclass` everywhere by its expansion (`@attr.s(auto_attribs=True, eq=False)`)? (For the reasons you mentioned)
But this is some great progress, thank you!

File: manim/mobject/geometry.py [hunk 2]
```diff
@@ -265,60 +271,74 @@ def stop_angle(self):
...
-                self.radius = np.linalg.norm(np.array(start) - np.array(center))
+                self.radius = np.linalg.norm(np.array(self.start) - np.array(center))
             else:
                 self.radius = math.inf
 
 
+@dclass
 class CurvedArrow(ArcBetweenPoints):
-    def __init__(self, start_point, end_point, **kwargs):
-        ArcBetweenPoints.__init__(self, start_point, end_point, **kwargs)
```
Comment (kilacoda) [inline manim/mobject/geometry.py:307]: Maybe change such `__init__` calls to `super().__init__` calls?

File: manim/animation/animation.py (hunk 1 above)
Comment (azarzadavila) [reply manim/animation/animation.py:19]: Yeah sorry I forgot to mention it. I removed it for the moment to test the current version with test_geometry. If I had kept the commit about animation (which is a progress of course) it would require more changes before being able to test it. I thought I could get the animation version from master and then perhaps cherry-pick later what was already done.

**Comment by kilacoda** [comment]
This is really good progress, thank you for it!

As you said, quite a few IDEs except PyCharm kinda suck when it comes to attrs. Thus, it'd be much appreciated if you provide a docstring with the args' information as well as a short description on the class for those who don't get class signatures properly. Here's an example of one I wrote for `Mobject`:

```
    """Mathematical Object - The most basic type of object meant to be displayed in a Scene.

    Args:
        name (Optional): The name given to the instance, defaults to the instance name.
        color (Union[str,Color]): The color of the instance. Defaults to `WHITE` i.e. #000000
        dim (int): The number of dimensions in which the instance's coordinates are defined. Defaults to 3. NOT meant to be changed.
        z_index (int): The precedence value among other objects in a Scene along the z-axis. Defaults to 0.

    """
```
(cc @PgBiel for his approval of this format)

Regarding the second inheritance issue, I was thinking (don't hate me for this) that we could keep the `digest_config` function in `__attrs_post_init__` so that we can modify such common properties without worrying about types and stuff. Better docstrings of all parent classes will be useful for this. Continuing your example:

```py
class Arc(TipableVMobject):
    color: typing.Union[str, Color] = WHITE

class Circle(Arc):
    CONFIG = {"color":RED} ##Type can be viewed via the docstring for parent class
```

I'll try looking into that first inheritance issue a bit later. Could you provide a code example for more clarity?

**Comment by PgBiel** [comment]
> As you said, quite a few IDEs except PyCharm kinda suck when it comes to attrs. Thus, it'd be much appreciated if you provide a docstring with the args' information as well as a short description on the class for those who don't get class signatures properly.

Yes, that should be specified as `Attributes` on class docstrings

> Here's an example of one I wrote for `Mobject`:
> ```
>     """Mathematical Object - The most basic type of object meant to be displayed in a Scene.
> 
>     Args:
>         name (Optional): The name given to the instance, defaults to the instance name.
>         color (Union[str,Color]): The color of the instance. Defaults to `WHITE` i.e. #000000
>         dim (int): The number of dimensions in which the instance's coordinates are defined. Defaults to 3. NOT meant to be changed.
>         z_index (int): The precedence value among other objects in a Scene along the z-axis. Defaults to 0.
> 
>     """
> ```
> 
> (cc @PgBiel for his approval of this format)

That's the Google format (except "Args", which would be "Attributes"), and is inconsistent with the rest of the codebase, which uses the Numpy format. Refer to `animations.py` for the format I use (or even just the doc guidelines - it's there).

> Regarding the second inheritance issue, I was thinking (don't hate me for this) that we could keep the `digest_config` function in `__attrs_post_init__` so that we can modify such common properties without worrying about types and stuff. Better docstrings of all parent classes will be useful for this. Continuing your example:

I'm pretty sure the whole point of this is to get rid of config (#7)... so no deal

One possible solution to this is to use the parent's `__annotations__`, i.e. `color: Arc.__annotations__["color"]` (we can write a utility for this so it doesn't look ugly), but idk if editors would support this stunt so

If we have to repeat it, then honestly so be it. (See my following comment)

**Comment by PgBiel** [comment]
The issue appointed as "MAIN ISSUE", actually, just shows how we need more consistency. The usage of attrs this way forces us to keep things consistent - if the same functionality changes on a child class, this can generate problems with ducktyping and whatnot (refer to Guido's comment on https://github.com/python/mypy/issues/1237).

And if we have to specify the types again on inherited classes, this only enforces consistency further, and if types change, we will be able to be sure that child classes adapt to it.

Of course, this could be made more convenient, but I'm yet to find out how

**Comment by azarzadavila** [comment]
Thank you for your comments. I'll then continue by propagating attrs in order to pass the tests and I'll add step by step the appropriate docstring.

**Comment by azarzadavila** [comment]
Just to clarify as asked the issue mentioned as "MAIN ISSUE". The following code gives an error with attrs : 
```python
@attr.s(auto_attribs=True)
class A:
    var1: int = 3  # keyword argument


@attr.s(auto_attribs=True)
class B(A):
    var2: int # positional argument with no default
```
The reason is that the constructor signature made by the decorator for `B` is
```python
def __init__(var1 = 3, var2)
```

**Comment by PgBiel** [comment]
Can you show a case in manim where this could happen?

**Comment by azarzadavila** [comment]
> Can you show a case in manim where this could happen?

Yes, for example if we look at ArcBetween points from master, the signature for initialization is :
```python
class ArcBetweenPoints(Arc):
    def __init__(self, start, end, angle=TAU / 4, radius=None, **kwargs):
```
which I would have changed to
```python
@attr.s(auto_attribs=True, eq=False)
class ArcBetweenPoints(Arc):
    start: np.ndarray
    end: np.ndarray
    angle: float = TAU / 4
    radius: typing.Optional[float] = None
```
This is not allowed because parent class (Arc, or even Mobject) has a keyword attribute.
This means there is no way to initialize by doing something like `ArcBetweenPoints(LEFT, RIGHT)`.
One could do `ArcBetweenPoints(start=LEFT, end=RIGHT)` or create a class method with signature
```python
    @classmethod
    def from_start_end(cls, start, end, **kwargs):
```
This would perhaps be more close to the philosophy attrs of only having attributes in the initialization. 
In this case, I don't think `start` and  `end` or kept as class attributes in the master version.

File: manim/mobject/geometry.py (hunk 2 above)
Comment (azarzadavila) [reply manim/mobject/geometry.py:307]: I sometimes feel like `super()` can behave a bit unexpected. I do get however that it seems more elegant due to the absence of reference to the parent class. I suggest I change this a bit latter when tests are passed and the whole PR feels more stable.

**Comment by azarzadavila** [comment]
The branch now passes all tests.

**Comment by leotrs** [comment]
@azarzadavila is it ready for review? If so, unmark it as a draft.

**Comment by azarzadavila** [comment]
> @azarzadavila is it ready for review? If so, unmark it as a draft.

I don't think it is ready to be merged. I would however appreciate comments on the work in progress as merging this would break current code used by manim users. More specifically some code using CONFIG would break and due to `attrs` design many code using initialization with positional arguments would also break.
For example, in the current state of the PR this
```python
class SceneTest(Scene):
    def construct(self):
        square = Square()
        circle = Circle()
        group = VGroup(square, circle)
```
should be replaced by
```python
class SceneTest(Scene):
    def construct(self):
        square = Square()
        circle = Circle()
        group = VGroup.from_vmobjects(square, circle)
```

File: manim/mobject/types/vectorized_mobject.py [hunk 3]
```diff
@@ -844,12 +850,48 @@ def get_subcurve(self, a, b):
...
+        kwargs
+
+        Returns
+        -------
+        VGroup
+        """
+        cls.check_vmobjects(*vmobjects)
+        instance = cls(**kwargs)
+        instance.add(*vmobjects)
+        return instance
```
Comment (leotrs) [inline manim/mobject/types/vectorized_mobject.py:894]: Question: Why is this necessary? Why doesn't `VGroup.add` perform the check itself?

**Comment by PgBiel** [comment]
Note that on the attrs branch we have also been adding typings and documentation - do you want to do it as well or delegate that work to us? (both options are ok)

File: manim/mobject/types/vectorized_mobject.py (hunk 3 above)
Comment (azarzadavila) [reply manim/mobject/types/vectorized_mobject.py:894]: It is not. Actually I think it could be more elegant to perform the check inside `VGroup.add` and then call the `super().add`. I'll change it :)

**Comment by azarzadavila** [comment]
> Note that on the attrs branch we have also been adding typings and documentation - do you want to do it as well or delegate that work to us? (both options are ok)

I plan on first keep propagating `attrs` (by first replacing the changes already done in Animation). So if you want to add typings and documentation now (or soon) it is fine by me, otherwise I can probably do some of the work later.

File: manim/animation/composition.py [hunk 4]
```diff
@@ -12,29 +14,27 @@
...
-        "group": None,
-    }
-
-    def __init__(self, *animations, **kwargs):
-        digest_config(self, kwargs)
-        self.animations = animations
+    # If None, this defaults to the sum of all
+    # internal animations
+    run_time: float = None
+    rate_func: tp.Any = linear
```
Comment (PgBiel) [inline manim/animation/composition.py:22]: ???
```suggestion
    rate_func: typing.Union[typing.Callable[[float, float], float], typing.Callable[[float], float]] \
        = linear
```

File: manim/animation/composition.py [hunk 5]
```diff
@@ -12,29 +14,27 @@
...
+    # If 0, all animations are played at once.
+    # If 1, all are played successively.
+    # If >0 and <1, they start at lagged times
+    # from one and other.
+    lag_ratio: float = 0
+    group: tp.Any = None
```
Comment (PgBiel) [inline manim/animation/composition.py:28]: ???
(Note: normally this would be Optional[Group], but since on init it is always set to a Group, then it's ok to have it as None)
```suggestion
    group: Group = None
```

File: manim/animation/composition.py [hunk 6]
```diff
@@ -12,29 +14,27 @@
...
+    animations: tp.List = attr.ib(default=attr.Factory(list))
```
Comment (PgBiel) [inline manim/animation/composition.py:29]: list of what? also just use this
```suggestion
    animations: tp.List[Animation] = attr.ib(factory=list)
```

File: manim/animation/creation.py [hunk 7]
```diff
@@ -24,17 +28,18 @@ def get_bounds(self, alpha):
...
+    lag_ratio: float = 1
 
     def get_bounds(self, alpha):
         return (0, alpha)
 
 
+@attr.s(auto_attribs=True, eq=False)
 class Uncreate(ShowCreation):
-    CONFIG = {"rate_func": lambda t: smooth(1 - t), "remover": True}
+    rate_func: tp.Any = lambda t: smooth(1 - t)
```
Comment (PgBiel) [inline manim/animation/creation.py:41]:
```suggestion
    rate_func: typing.Union[typing.Callable[[float, float], float], typing.Callable[[float], float]] \
        = lambda t: smooth(1 - t)
```

File: manim/animation/indication.py [hunk 8]
```diff
@@ -51,12 +55,14 @@ def create_starting_mobject(self):
...
 
 
+@attr.s(auto_attribs=True, eq=False)
 class Indicate(Transform):
-    CONFIG = {
-        "rate_func": there_and_back,
-        "scale_factor": 1.2,
-        "color": YELLOW,
-    }
+    rate_func: tp.Any = there_and_back
```
Comment (PgBiel) [inline manim/animation/indication.py:60]:
```suggestion
    rate_func: typing.Union[typing.Callable[[float, float], float], typing.Callable[[float], float]] \
        = there_and_back
```

File: manim/animation/indication.py [hunk 9]
```diff
@@ -65,29 +71,27 @@ def create_target(self):
...
-    }
-
-    def __init__(self, point, color=YELLOW, **kwargs):
-        self.point = point
-        self.color = color
-        digest_config(self, kwargs)
+    line_length: float = 0.2
+    num_lines: int = 12
+    flash_radius: float = 0.3
+    line_stroke_width: float = 3
```
Comment (PgBiel) [inline manim/animation/indication.py:79]: this is clearly not a float
```suggestion
    line_stroke_width: tp.Union[int, float] = 3
```

File: manim/animation/creation.py [hunk 10]
```diff
@@ -24,17 +28,18 @@ def get_bounds(self, alpha):
         raise Exception("Not Implemented")
 
 
+@attr.s(auto_attribs=True, eq=False)
 class ShowCreation(ShowPartial):
-    CONFIG = {
-        "lag_ratio": 1,
-    }
+    lag_ratio: float = 1
```
Comment (PgBiel) [inline manim/animation/creation.py:33]:
```suggestion
    lag_ratio: tp.Union[int, float] = 1
```

File: manim/animation/composition.py [hunk 11]
```diff
@@ -96,10 +96,12 @@ def interpolate(self, alpha):
             anim.interpolate(sub_alpha)
 
 
+@attr.s(auto_attribs=True, eq=False)
 class Succession(AnimationGroup):
-    CONFIG = {
-        "lag_ratio": 1,
-    }
+    lag_ratio: float = 1
```
Comment (PgBiel) [inline manim/animation/composition.py:101]:
```suggestion
    lag_ratio: tp.Union[int, float] = 1
```

File: manim/animation/composition.py [hunk 12]
```diff
@@ -12,29 +14,27 @@
...
-        self.animations = animations
+    # If None, this defaults to the sum of all
+    # internal animations
+    run_time: float = None
+    rate_func: tp.Any = linear
+    # If 0, all animations are played at once.
+    # If 1, all are played successively.
+    # If >0 and <1, they start at lagged times
+    # from one and other.
+    lag_ratio: float = 0
```
Comment (PgBiel) [inline manim/animation/composition.py:27]:
```suggestion
    lag_ratio: tp.Union[int, float] = 0
```

File: manim/animation/fading.py [hunk 13]
```diff
@@ -42,16 +46,13 @@ def create_starting_mobject(self):
...
 
 
+@attr.s(auto_attribs=True, eq=False)
 class FadeInFrom(Transform):
-    CONFIG = {
-        "direction": DOWN,
-        "lag_ratio": DEFAULT_ANIMATION_LAG_RATIO,
-    }
+    direction: np.ndarray = DOWN
+    lag_ratio: float = DEFAULT_ANIMATION_LAG_RATIO
```
Comment (PgBiel) [inline manim/animation/fading.py:52]:
```suggestion
    lag_ratio: tp.Union[int, float] = DEFAULT_ANIMATION_LAG_RATIO
```

File: manim/animation/indication.py [hunk 14]
```diff
@@ -142,50 +144,50 @@ def finish(self):
...
 
 
+@attr.s(auto_attribs=True, eq=False)
 class ShowCreationThenDestruction(ShowPassingFlash):
-    CONFIG = {
-        "time_width": 2.0,
-        "run_time": 1,
-    }
+    time_width: float = 2.0
+    run_time: float = 1
```
Comment (PgBiel) [inline manim/animation/indication.py:150]:
```suggestion
    run_time: tp.Union[int, float] = 1
```

File: manim/animation/indication.py [hunk 15]
```diff
@@ -142,50 +144,50 @@ def finish(self):
...
+@attr.s(auto_attribs=True, eq=False)
 class AnimationOnSurroundingRectangle(AnimationGroup):
-    CONFIG = {
-        "surrounding_rectangle_config": {},
-        # Function which takes in a rectangle, and spits
-        # out some animation.  Could be some animation class,
-        # could be something more
-        "rect_animation": Animation,
-    }
+    surrounding_rectangle_config: tp.Dict = attr.ib(default=attr.Factory(dict))
```
Comment (PgBiel) [inline manim/animation/indication.py:164]: well, this should be removed as soon as we finish attrs, or maybe even during config removal, as this would go to CONFIG, which won't exist anymore. Rather, the user should specify their own SurroundingRectangle instance.
```suggestion
    surrounding_rectangle_config: dict = attr.ib(factory=dict)  
```

File: manim/animation/indication.py [hunk 16]
```diff
@@ -142,50 +144,50 @@ def finish(self):
...
+    # Function which takes in a rectangle, and spits
+    # out some animation.  Could be some animation class,
+    # could be something more
+    rect_animation: type = Animation
+    kwargs_animations: tp.Dict = attr.ib(default=attr.Factory(dict))
```
Comment (PgBiel) [inline manim/animation/indication.py:169]: note that we generally use tp.Dict to specify the types of the dict's keys and values: tp.Dict[K, V]
```suggestion
    kwargs_animations: dict = attr.ib(default=attr.Factory(dict))
```

File: manim/animation/transform.py [hunk 17]
```diff
@@ -15,17 +17,15 @@
...
 class Transform(Animation):
-    CONFIG = {
-        "path_arc": 0,
-        "path_arc_axis": OUT,
-        "path_func": None,
-        "replace_mobject_with_target_in_scene": False,
-    }
+    path_arc: float = 0
+    path_arc_axis: np.ndarray = OUT
+    path_func: tp.Any = None
```
Comment (PgBiel) [inline manim/animation/transform.py:24]: it can't be any as it expects a function, so at least a Callable. We gotta work on inferring the expected function signature though  (TODO)
```suggestion
    path_func: tp.Callable = None
```

File: manim/mobject/coordinate_systems.py [hunk 18]
```diff
@@ -119,27 +120,26 @@ def input_to_graph_point(self, x, graph):
                 return None
 
 
+@attr.s(auto_attribs=True, eq=False)
 class Axes(VGroup, CoordinateSystem):
-    CONFIG = {
-        "axis_config": {
+    axis_config: tp.Dict = {
```
Comment (PgBiel) [inline manim/mobject/coordinate_systems.py:125]: same comment about removing CONFIG later.
```suggestion
    axis_config: dict = {
```

File: manim/mobject/coordinate_systems.py [hunk 19]
```diff
@@ -119,27 +120,26 @@ def input_to_graph_point(self, x, graph):
...
             "color": LIGHT_GREY,
             "include_tip": True,
             "exclude_zero_from_default_numbers": True,
-        },
-        "x_axis_config": {},
-        "y_axis_config": {"label_direction": LEFT,},
-        "center_point": ORIGIN,
     }
+    x_axis_config: tp.Dict = attr.ib(default=attr.Factory(dict))
```
Comment (PgBiel) [inline manim/mobject/coordinate_systems.py:130]: Same comment about removing CONFIG later.
```suggestion
    x_axis_config: dict = attr.ib(factory=dict)
```

File: manim/mobject/coordinate_systems.py [hunk 20]
```diff
@@ -119,27 +120,26 @@ def input_to_graph_point(self, x, graph):
...
+    y_axis_config: tp.Dict = attr.ib(default=attr.Factory(lambda: {"label_direction": LEFT, }))
```
Comment (PgBiel) [inline manim/mobject/coordinate_systems.py:131]: ^
```suggestion
    y_axis_config: dict = attr.ib(factory=lambda: {"label_direction": LEFT, })
```

File: manim/mobject/coordinate_systems.py [hunk 21]
```diff
@@ -176,31 +176,30 @@ def get_coordinate_labels(self, x_vals=None, y_vals=None):
...
-    }
-
-    def __init__(self, **kwargs):
-        Axes.__init__(self, **kwargs)
+    dimension: int = 3
+    x_min: float = -5.5
+    x_max: float = 5.5
+    y_min: float = -5.5
+    y_max: float = 5.5
+    z_axis_config: tp.Dict = attr.ib(default=attr.Factory(dict))
```
Comment (PgBiel) [inline manim/mobject/coordinate_systems.py:194]:
```suggestion
    z_axis_config: dict = attr.ib(factory=dict)
```

File: manim/mobject/functions.py [hunk 22]
```diff
@@ -3,22 +3,22 @@
...
-    CONFIG = {
-        "t_min": 0,
-        "t_max": 1,
-        "step_size": 0.01,  # Use "auto" (lowercase) for automatic step size
-        "dt": 1e-8,
-        # TODO, be smarter about figuring these out?
-        "discontinuities": [],
-    }
+    t_min: float = 0
+    t_max: float = 1
```
Comment (PgBiel) [inline manim/mobject/functions.py:13]:
```suggestion
    t_min: tp.Union[int, float] = 0
    t_max: tp.Union[int, float] = 1
```

File: manim/mobject/functions.py [hunk 23]
```diff
@@ -3,22 +3,22 @@
...
+    step_size: tp.Union[str, float] = 0.01  # Use "auto" (lowercase) for automatic step size
```
Comment (PgBiel) [inline manim/mobject/functions.py:14]: Comment: note that, if on python 3.8+, `str` can be replaced by `tp.Literal["auto"]`. I suppose we can add a compatibility check: (before the class)
`auto_lit = tp.Literal["auto"] if hasattr(tp, "Literal") else str`
then use it here, so that people in python 3.8+ can have proper typings
but that's just an idea

File: manim/mobject/functions.py [hunk 24]
```diff
@@ -3,22 +3,22 @@
...
+    dt: float = 1e-8
+    # TODO, be smarter about figuring these out?
+    discontinuities: tp.List = []
```
Comment (PgBiel) [inline manim/mobject/functions.py:17]: list of what?

File: manim/mobject/geometry.py [hunk 25]
```diff
@@ -46,12 +57,12 @@ class TipableVMobject(VMobject):
...
 
-    CONFIG = {
-        "tip_length": DEFAULT_ARROW_TIP_LENGTH,
-        # TODO
-        "normal_vector": OUT,
-        "tip_style": {"fill_opacity": 1, "stroke_width": 0,},
-    }
+    tip_length: float = DEFAULT_ARROW_TIP_LENGTH
+    normal_vector: np.ndarray = OUT
+    tip_style: dict = {"fill_opacity": 1, "stroke_width": 0}
```
Comment (PgBiel) [inline manim/mobject/geometry.py:62]:
```suggestion
    tip_style: dict = attr.ib(factory=lambda: {"fill_opacity": 1, "stroke_width": 0})
```

File: manim/mobject/geometry.py [hunk 26]
```diff
@@ -192,22 +203,23 @@ def get_length(self):
...
-    def __init__(self, start_angle=0, angle=TAU / 4, **kwargs):
-        self.start_angle = start_angle
-        self.angle = angle
-        self._failed_to_get_center = False
-        VMobject.__init__(self, **kwargs)
+    radius: float = 1.0
+    num_components: int = 9
+    anchors_span_full_range: bool = True
+    arc_center: np.ndarray = ORIGIN
+    start_angle: float = 0
```
Comment (PgBiel) [inline manim/mobject/geometry.py:212]:
```suggestion
    start_angle: tp.Union[int, float] = 0
```

File: manim/mobject/geometry.py [hunk 27]
```diff
@@ -265,60 +277,74 @@ def stop_angle(self):
...
     """
     Inherits from Arc and additionally takes 2 points between which the arc is spanned.
     """
 
-    def __init__(self, start, end, angle=TAU / 4, radius=None, **kwargs):
-        if radius is not None:
-            self.radius = radius
-            if radius < 0:
+    start: tp.Any = None
+    end: tp.Any = None
```
Comment (PgBiel) [inline manim/mobject/geometry.py:287]: iterable, as it just does `np.array(start)` and `np.array(end)`
Although I guess we could simply enforce the usage of `np.ndarray`

```suggestion
    start: tp.Iterable[tp.Union[int, float]] = None
    end: tp.Iterable[tp.Union[int, float]] = None
```
```suggestion
    start: np.ndarray = None
    end: np.ndarray = None
```

File: manim/mobject/geometry.py [hunk 28]
```diff
@@ -265,60 +277,74 @@ def stop_angle(self):
...
             else:
                 self.radius = math.inf
 
 
+@attr.s(auto_attribs=True, eq=False)
 class CurvedArrow(ArcBetweenPoints):
-    def __init__(self, start_point, end_point, **kwargs):
-        ArcBetweenPoints.__init__(self, start_point, end_point, **kwargs)
+    start_point: tp.Any = None
+    end_point: tp.Any = None
```
Comment (PgBiel) [inline manim/mobject/geometry.py:322]:
```suggestion
    start_point: tp.Iterable[tp.Union[int, float]] = None
    end_point: tp.Iterable[tp.Union[int, float]] = None
```

File: manim/mobject/geometry.py [hunk 29]
```diff
@@ -336,43 +362,50 @@ def point_at_angle(self, angle):
...
+@attr.s(auto_attribs=True, eq=False)
 class Dot(Circle):
-    CONFIG = {
-        "radius": DEFAULT_DOT_RADIUS,
-        "stroke_width": 0,
-        "fill_opacity": 1.0,
-        "color": WHITE,
-    }
+    radius: float = DEFAULT_DOT_RADIUS
+    stroke_width: float = 0
```
Comment (PgBiel) [inline manim/mobject/geometry.py:368]: that or change to 0.0
```suggestion
    stroke_width: tp.Union[int, float] = 0
```
```suggestion
    stroke_width: float = 0.0
```

File: manim/mobject/geometry.py [hunk 30]
```diff
@@ -336,43 +362,50 @@ def point_at_angle(self, angle):
...
-        "angle": TAU / 4,
-        "start_angle": 0,
-        "fill_opacity": 1,
-        "stroke_width": 0,
-        "color": WHITE,
-    }
+    inner_radius: float = 1.0
+    outer_radius: float = 2.0
+    angle: float = TAU / 4
+    start_angle: float = 0
```
Comment (PgBiel) [inline manim/mobject/geometry.py:402]: that or change to 0.0
```suggestion
    start_angle: tp.Union[int, float] = 0
```
```suggestion
    start_angle: float = 0.0
```

File: manim/mobject/geometry.py [hunk 31]
```diff
@@ -415,16 +455,17 @@ def generate_points(self):
         self.shift(self.arc_center)
 
 
+@attr.s(auto_attribs=True, eq=False)
 class Line(TipableVMobject):
-    CONFIG = {
-        "buff": 0,
-        "path_arc": None,  # angle of arc specified here
-    }
+    buff: float = 0
```
Comment (PgBiel) [inline manim/mobject/geometry.py:460]:
```suggestion
    buff: tp.Union[int, float] = 0
```
```suggestion
    buff: float = 0.0
```

File: manim/mobject/geometry.py [hunk 32]
```diff
@@ -560,46 +600,46 @@ def get_last_handle(self):
...
 
 
+@attr.s(auto_attribs=True, eq=False)
 class Elbow(VMobject):
-    CONFIG = {
-        "width": 0.2,
-        "angle": 0,
-    }
+    width: float = 0.2
+    angle: float = 0
```
Comment (PgBiel) [inline manim/mobject/geometry.py:624]:
```suggestion
    angle: tp.Union[int, float] = 0
```
```suggestion
    angle: float = 0.0
```

File: manim/mobject/geometry.py [hunk 33]
```diff
@@ -560,46 +600,46 @@ def get_last_handle(self):
...
-    }
-
-    def __init__(self, *args, **kwargs):
-        Line.__init__(self, *args, **kwargs)
-        # TODO, should this be affected when
-        # Arrow.set_stroke is called?
+    stroke_width: float = 6.0
+    buff: float = MED_SMALL_BUFF
+    max_tip_length_to_length_ratio: float = 0.25
+    max_stroke_width_to_length_ratio: float = 5
```
Comment (PgBiel) [inline manim/mobject/geometry.py:638]:
```suggestion
    max_stroke_width_to_length_ratio: tp.Union[int, float] = 5
```
```suggestion
    max_stroke_width_to_length_ratio: float = 5.0
```

**Comment by leotrs** [comment]
This PR is **way** too large to review effectively. In the future we should make smaller, incremental, self-contained PRs in general

**Comment by PgBiel** [comment]
> In the future we should make smaller, incremental, self-contained PRs in general

In fact, I think we should start immediately. Our plan was to split our work in attrs; I proposed doing one PR per folder, while EulerTour proposed a further division. We can't work with a PR for everything. So I propose we split this here and now

**Comment by leotrs** [comment]
Yep, let's do it. 251 files in a single PR is unmanageable I think.

**Comment by azarzadavila** [comment]
Okey, sorry I was not sure how to proceed with this PR. I also thought merging master would look nicer than just show all commits since beginning of the PR. I wanted to be able to pass all tests before putting this ready to review. How were you thinking of proceeding? Perhaps it would be interesting to keep a separate branch and applying multiple PRs to that branch until it has at least everything to pass the tests. For example, I think it was necessary once Mobject pass to attrs every Mobject subclass should also use attrs to avoid any errors.

**Comment by leotrs** [comment]
Whoever is working on the attrs branch should periodically merge master into attrs (NOT the other way around). And attrs should 
be merged into master frequently once each file or folder is done.

**Comment by leotrs** [comment]
I think the way forward here is to delete this PR, merge master into attrs and resolve merge conflicts, and PR from attrs again, only a few files at a time if possible.

**Comment by PgBiel** [comment]
> Okey, sorry I was not sure how to proceed with this PR.
It's perfectly fine! Sorry if it sounded like some sort of rant or something lol, I should have expressed myself better... Just, it needed to be mentioned

> I also thought merging master would look nicer than just show all commits since beginning of the PR.
We can prob do this in the attrs branch as leo proposed above

> I wanted to be able to pass all tests before putting this ready to review.
Well, that's certainly an issue: by splitting PRs, tests won't pass just yet. But once we build the attrs branch slowly (see below), at the end, we will be able to verify tests with the "big PR".

> How were you thinking of proceeding?
(see below)

> Perhaps it would be interesting to keep a separate branch and applying multiple PRs to that branch until it has at least everything to pass the tests.
The idea is similar: we'd go merging PRs until we finish the job, then we'd make one _master_ PR (haha get it) (ok sorry that was bad) to concretize the changes

> For example, I think it was necessary once Mobject pass to attrs every Mobject subclass should also use attrs to avoid any errors.
Doing this would generate necessarily big PRs. We can't depend on the inheritance tree. There are better options, such as per folder or idk 2 per folder, gotta see that

**Comment by PgBiel** [comment]
> I think the way forward here is to delete this PR, merge master into attrs and resolve merge conflicts, and PR from attrs again, only a few files at a time if possible.

sure thing bud

**Comment by leotrs** [comment]
@azarzadavila @PgBiel 

As per the comments here, it doesn't seem like we are going to move forward with this PR, right? It is my understanding that this will be broken down into several smaller PRs. If that's the case, please close this one, else please update on the status.

**Comment by leotrs** [comment]
I'm closing this for lack of activity.
